File: clojure_hacking/day6.clj
(def input [14 0 15 12 11 11 3 5 1 6 8 4 9 1 8 4])

; these two could probably be more functional, but I want to do it in one pass :P
(defn- find-hi [nums]
       (loop [num-max Long/MIN_VALUE
              max-idx 0
              cur-idx 0]
             (cond
               (>= cur-idx (count nums)) max-idx
               (> (nums cur-idx) num-max) (recur (nums cur-idx) cur-idx (inc cur-idx))
               :else (recur num-max max-idx (inc cur-idx)))))

(defn- redistribute [nums idx]
       (loop [res (assoc! (transient nums) idx 0)
              redist (nums idx)
              cur-idx (mod (inc idx) (count nums))]
             (if (zero? redist)
               (persistent! res)
               (recur (assoc! res cur-idx (inc (res cur-idx)))
                      (dec redist)
                      (mod (inc cur-idx) (count nums))))))

(loop [seen? {input 0}
       cur input
       iters 0]
      (let [hi (find-hi cur)
            rd (redistribute cur hi)]
           (if (seen? rd)
             (do (println "part 1:" (inc iters))
                 (println "part 2:" (- (inc iters) (seen? rd))))
             (recur (assoc seen? rd (inc iters)) rd (inc iters)))))

File: src/day6.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::BufRead;
use std::io::BufReader;

pub fn run() {
    let mut input = vec![14, 0, 15, 12, 11, 11, 3, 5, 1, 6, 8, 4, 9, 1, 8, 4];

    let mut seen = HashMap::new();
    seen.insert(input.clone(), 0);

    let mut iters = 0;
    loop {
        let hi = find_highest(&input);
        redistribute(&mut input, hi);
        iters += 1;
        if seen.contains_key(&input) {
            println!("part 1: {}", iters);
            println!("part 2: {}", iters - seen[&input]);
            break;
        } else {
            seen.insert(input.clone(), iters);
        }
    }
}

fn redistribute(nums: &mut [u32], start_from: usize) {
    let mut redist = nums[start_from];
    nums[start_from] = 0;
    let mut idx = (start_from + 1) % nums.len();
    while redist > 0 {
        nums[idx] += 1;
        redist -= 1;
        idx = (idx + 1) % nums.len();
    }
}

fn find_highest(nums: &[u32]) -> usize {
    let mut max = u32::min_value();
    let mut max_idx = 0;

    for (idx, &i) in nums.iter().enumerate() {
        if i > max {
            max = i;
            max_idx = idx;
        }
    }

    max_idx
}